### ISSUE TITLE
optionalDependencies should be installed with lerna bootstrap 

### ISSUE THREAD

**Issue by robmcguinness** [lerna/lerna#1350]
## v3.0.0-beta.9

`lerna bootstrap` does not install `optionalDependencies` for packages managed by Lerna. I use optional dependencies to install developer tools like `sqlite` and `node-keytar` locally while preventing their installation on CI (Jenkins) severs using `--no-optional` flag.

## Expected Behavior
`lerna bootstrap` should install `optionalDependencies`.

## Current Behavior
Only `dependencies`, `devDependencies` and `peerDependencies` are accounted for when bootstrapping while `optionaDependencies` are ignored.

## Possible Solution
- Add `optionalDependencies` to https://github.com/lerna/lerna/blob/master/core/package/index.js#L58-L7

```js
 Object.defineProperties(this, {
      // ...
      dependencies: {
        get() {
          return pkg.dependencies;
        },
      },
      devDependencies: {
        get() {
          return pkg.devDependencies;
        },
      },
      peerDependencies: {
        get() {
          return pkg.peerDependencies;
        },
      },
      optionalDependencies: {
        get() {
          return pkg.optionalDependencies;
        },
      }, 
     // ...
    });
```

- Install `optionalDependencies` when `allDependencies` is true https://github.com/lerna/lerna/blob/master/core/package-graph/index.js#L67-L69

```js
this.forEach((currentNode, currentName) => {
    const graphDependencies =
        graphType === 'dependencies'
            ? Object.assign({}, currentNode.pkg.dependencies)
            : Object.assign(
                {},
                currentNode.pkg.devDependencies,
                currentNode.pkg.optionalDependencies,
                currentNode.pkg.dependencies
```

## Steps to Reproduce (for bugs)

```json
{
  "lerna": "3.0.0-beta.9",
  "packages": ["packages/*"],
  "version": "0.50.0"
}
```

## Your Environment
```
System:
    OS: macOS Sierra 10.12.6
    CPU: x64 Intel(R) Core(TM) i7-7820HQ CPU @ 2.90GHz
    Free Memory: 1.16 GB
    Total Memory: 16.00 GB
    Shell: /bin/bash - 3.2.57
  Binaries:
    Node: 8.9.4
    Yarn: 1.5.1
    npm: 5.8.0
```

Thoughts?

**Comment by evocateur** [comment]
#121 #191 #1136

Adding `optionalDependencies` to the package graph (to ensure the `externalDependencies` are complete) means we _also_ need to add support for leaf packages linked via `optionalDependencies`. Sounds weird, but it's the least-worst option, I suppose.

**Comment by evocateur** [comment]
Implemented in b73e19df
